_posts/2017-03-30-storage-formats-quickstep.markdown

```diff
@@ -0,0 +1,140 @@
+
+<h1>Storage Formats 101</h1>
+A storage format refers to how the data in a table is laid out in the memory. Let's consider a toy relational schema.
+
+{% highlight sql %}
+CREATE TABLE employee (ID INT, Age INT, Name VARCHAR(8));
+{% endhighlight %}
+
+A tuple in the employee table consists of three attributes, id (an integer), age (an integer) and the name (character of 8 bytes fixed length). Let's also populate our table with few tuples. For simplicity, let us assume that the name is a fixed length attribute; however, in general it could be variable length. 
+
+{% highlight sql %}
+INSERT into employee values (101, 30, Jennifer);
+INSERT into employee values (102, 25, Jim);
+INSERT into employee values (103, 35, David);
+{% endhighlight %}
+
+Conceptually, our table looks as follows:
+
+| ID 	| Age 	|   Name   	|
+|:--:	|:---:	|:--------:	|
+|  101 	|  30 	| Jennifer 	|
+|  102 	|  25 	|    Jim   	|
+|  103 	|  35 	|   David  	|
+
+Next, let us dig into understanding the different ways in which this table can be stored in memory. 
+
+<h2>Row Store</h2>
+
+The first storage format we will look at is *row store*, which is a popular format, and one that is intuitive too.
+
+The figure below represents how a row store format for our toy example looks like in memory. The offsets indicate how far is the memory location from the beginning of the block. 
+
+![Row Store](../assets/storage-format-row-store.jpg)
+
+In the row store format, a tuple is stored in the table definition order. In memory, the ID (101) of the first tuple is stored at some address, and it is followed by the age field for that tuple (30), and then the name field ("Jennifer"). The second tuple is stored in memory sequentially after the first tuple, and so on.
+
+<h2>Column Store</h2>
+
+Let's now take a look at the *column store* format. The figure below depicts a column store layout for our toy example. 
+
+![Column Store](../assets/storage-format-column-store.jpg)
+
+In the column store format, the values from the same column are stored together. The order in which the values from a column are stored remains the same for all the columns. Notice the ID values 101, 102, 103 are stored contiguously. The corresponding Age fields 30, 25 and 35 are stored contiguously and likewise for the Name column.
```

Comment:
I think the visual is good, but it would only be true for a block that was small enough to fit exactly 3 tuples. Otherwise the block calculates the maximum number of tuples that can fit in the given space. Then it divides sequential segments which the length of `max_tuples * sizeof(attribute_type)` so that there is empty space between attribute regions. If you don't want to change the vis, I could see an argument for keeping it simple, as well.